db-engine/src/main/java/com/splicemachine/db/iapi/sql/compile/Optimizable.java

```diff
@@ -470,4 +470,22 @@ default void setHasJoinPredicatePushedDownFromOuter(boolean value) {
      * the first index column.
      */
     default boolean indexPrefixIteratorAllowed(AccessPath accessPath) { return false; }
+
+    /**
+     * If true, this optimizable must be the outer table of a join, and
+     * join plans which use it as the inner table will not be feasible.
+     * For a given binary join involving an <code>Optimizable</code> marked as
+     * <code>outerTableOnly</code>, the optimizer will only allow that
+     * <code>Optimizable</code> to participate as the outer table of the join.
+     */
+    default boolean outerTableOnly() { return false; }
```

Comment:
This setting is only intended to be used in the simple case of joining one table with another, like in bindAndOptimizeUnionedIndexScansPath.  Complex join scenarios are not targeted or considered.  To clarify this, I have added the following comment to the above:

     * This will only return true in limited scenarios when a table may
     * only join with a single other table and no public interface to
     * mark an optimizable as outerTableOnly is provided, on purpose,
     * as this is an internal limited-use mechanism.

Also, I have made setOuterTableOnly protected, and added the following comment:
    // This is a special-purpose method for tagging an optimizable to be used
    // as the outer table of a join in very limited scenarios, usually when
    // we are planning only a single join between it and another table.
    // Please do not use this for queries of all forms, and only
    // use it if you have a targeted, specific and limited use case.
    protected void setOuterTableOnly(boolean outerTableOnly) {

The only Optimizable that is planned outside of bindAndOptimizeUnionedIndexScansPath with other Optimizables is the uisRowIdJoinBackToBaseTableResultSet.  The flag is set in initializeUnionedIndexScanAccessPath.  I'm thinking this may not actually be required.  So, to limit the scope of usage of this setting I will remove the call to setOuterTableOnly in this method.

Reply:
Thanks. I think it really makes sense to document it and prevent it from being misused. Change is good.